### ISSUE TITLE
Not working

### ISSUE THREAD

**Issue by JDeepD** [subnut/capsctrl#1]
> \>>> ./example.sh

>Something is wrong with your format string.
Something is wrong with your format string.
Something is wrong with your format string.
Something is wrong with your format string.
Something is wrong with your format string.
Something is wrong with your format string.
Failed to initialize libevdev device: Inappropriate ioctl for device
Failed to initialize libevdev device: Inappropriate ioctl for device
Something is wrong with your format string.

I also tried running it in background using `capsctrl &`. Did not work
>[1] 110080
Usage: capsctrl <device>
<device>	Path to device inside /dev/input
[1]  + exit 1     capsctrl

**Comment by subnut** [comment]
Give me the output of 
```
ls -l /dev/input/by-path
```

**Comment by JDeepD** [comment]
> Give me the output of
> 
> ```
> ls -l /dev/input/by-path

 ```
total 0
lrwxrwxrwx 1 root root 10 Nov 19 17:14 pci-0000:00:14.0-usb-0:4:1.1-event-kbd -> ../event16
lrwxrwxrwx 1 root root 10 Nov 19 17:14 pci-0000:00:14.0-usb-0:4:1.1-event-mouse -> ../event17
lrwxrwxrwx 1 root root  9 Nov 19 17:14 pci-0000:00:14.0-usb-0:4:1.1-mouse -> ../mouse2
lrwxrwxrwx 1 root root 10 Nov 19 17:14 pci-0000:00:14.0-usb-0:6:1.0-event -> ../event18
lrwxrwxrwx 1 root root  9 Nov 19 17:14 pci-0000:00:15.1-platform-i2c_designware.1-event-mouse -> ../event7
lrwxrwxrwx 1 root root  9 Nov 19 17:14 pci-0000:00:15.1-platform-i2c_designware.1-mouse -> ../mouse0
lrwxrwxrwx 1 root root  9 Nov 19 17:14 platform-asus-nb-wmi-event -> ../event4
lrwxrwxrwx 1 root root  9 Nov 19 17:14 platform-i8042-serio-0-event-kbd -> ../event3
lrwxrwxrwx 1 root root  9 Nov 19 17:14 platform-pcspkr-event-spkr -> ../event6
```

**Comment by subnut** [comment]
put this in a script and run it -
```sh
#!/usr/bin/bash
sleep 1
echo Forking...
capsctrl '/dev/input/by-path/pci-0000:00:14.0-usb-0:4:1.1-event-kbd' &!
capsctrl '/dev/input/by-path/platform-i8042-serio-0-event-kbd' &!
echo Done
```
Tell me if it works

---
The problem seems to be due to `inotifywait` not being able to understand the format string?
What version of `inotify-tools` are you using? (The version can be found at the very end of `man inotifywait`)

**Comment by JDeepD** [comment]
> The problem seems to be due to `inotifywait` not being able to understand the format string? What version of `inotify-tools` are you using? (The version can be found at the very end of `man inotifywait`)

`inotifywait 3.21.9.6`

> Tell me if it works

Should I put the above lines inside the `while` loop or just at the end?

**Comment by JDeepD** [comment]
Output with above code :

>Forking...
Done
Failed to grab device: Device or resource busy
Failed to grab device: Device or resource busy
Something is wrong with your format string.
Failed to initialize libevdev device: Inappropriate ioctl for device

**Comment by subnut** [comment]
> Should I put the above lines inside the `while` loop or just at the end?

Just make a separate script, `chmod +x` it and execute it.

**Comment by JDeepD** [comment]
```
Forking...
Done
Failed to grab device: Device or resource busy
Failed to grab device: Device or resource busy
```

**Comment by subnut** [comment]
```sh
sudo pkill capsctrl
```
and then again run the above script and give me the output

**Comment by JDeepD** [comment]
Great. Its working now 👍🏻

**Comment by JDeepD** [comment]
There is a bug though.

**Comment by JDeepD** [comment]
https://user-images.githubusercontent.com/64089730/142738076-c402360c-0e79-4c83-af28-89640b76e6cf.mp4

For some reason, Enter was pressed forever after the script was run.

```sh
#!/bin/bash
ps cax | grep capsctrl > /dev/null
if [ $? -eq 0 ]; then
  echo "Capsctrl is already running."
  exit 1
fi

echo Forking...
capsctrl '/dev/input/by-path/pci-0000:00:14.0-usb-0:4:1.1-event-kbd' &!
capsctrl '/dev/input/by-path/platform-i8042-serio-0-event-kbd' &!
echo Done

```

**Comment by subnut** [comment]
> For some reason, Enter was pressed forever after the script was run.

Yeah. Actually, when you press Enter to run the command, your terminal observes it as a `KeyDown` event. But by the time you lift your Enter key, `capsctrl` has already taken control of the keyboard and your terminal never recieves the `KeyUp` event. Thereby, it assumes that the Enter key has been held down.

It's the same as attaching a USB keyboard, pressing Enter on it and removing the keyboard from the USB port without releasing the Enter key.

The `sleep 1` in my script was intended to solve that. Try changing it to `sleep 3`.  
Basically the intention is to delay executing `capsctrl` for a sufficient time that the user is able to lift the Enter key after executing the script.

So if I take your script for example...
```diff
 fi
 
+sleep 3
 echo Forking...
 capsctrl '/dev/input/by-path/pci-0000:00:14.0-usb-0:4:1.1-event-kbd' &!
```
---
Your suggested script will run all right, but it won't work if a new keyboard is added. I was using `inotifywait` for that express purpose.

BTW u can use `pgrep` instead of `ps cax | grep`

**Comment by subnut** [comment]
The exact command for `inotifywait` is -
```bash
inotifywait -q -e create --include '.*-event-kbd' --format %f --no-newline /dev/input/by-path
```
try understanding what it does and try debugging it yourself :)

**Comment by JDeepD** [comment]
> > For some reason, Enter was pressed forever after the script was run.
> 
> Yeah. Actually, when you press Enter to run the command, your terminal observes it as a `KeyDown` event. But by the time you lift your Enter key, `capsctrl` has already taken control of the keyboard and your terminal never recieves the `KeyUp` event. Thereby, it assumes that the Enter key has been held down.
> 
> It's the same as attaching a USB keyboard, pressing Enter on it and removing the keyboard from the USB port without releasing the Enter key.
> 
> The `sleep 1` in my script was intended to solve that. Try changing it to `sleep 3`. Basically the intention is to delay executing `capsctrl` for a sufficient time that the user is able to lift the Enter key after executing the script.
> 
> So if I take your script for example...
> 
> ```diff
>  fi
>  
> +sleep 3
>  echo Forking...
>  capsctrl '/dev/input/by-path/pci-0000:00:14.0-usb-0:4:1.1-event-kbd' &!
> ```
> 
> Your suggested script will run all right, but it won't work if a new keyboard is added. I was using `inotifywait` for that express purpose.
> 
> BTW u can use `pgrep` instead of `ps cax | grep`

oh ok
